File: examples/check-network-errors.mjs
/**
 * Webcompat repro pattern: network request failures (CORS, 4xx, 5xx, fetch errors)
 *
 * Use this when you observed: a failed fetch in DevTools, a CORS error in the
 * console, missing content that should have loaded from an API, or a red request
 * in the network panel.
 *
 * Adjust TARGET_URL and optionally FILTER_URL to narrow results to a specific
 * endpoint you identified.
 * Run: ff-bidi-run examples/check-network-errors.mjs --firefox /path/to/firefox --headless
 */
import { startSession } from 'ff-test-firefox-bidi-client';

const TARGET_URL = 'https://example.com';

// Optional: only report failures for URLs containing this string.
// Set to '' to report all failures.
const FILTER_URL = '';

const session = await startSession();
const { page, connection } = session;

const failures = [];

connection.onEvent('network.fetchError', params => {
  const url = params.request.url;
  if (!FILTER_URL || url.includes(FILTER_URL)) {
    failures.push({ type: 'fetchError', url, error: params.errorText });
  }
});

connection.onEvent('network.responseCompleted', params => {
  const url = params.request.url;
  const status = params.response.status;
  if (status >= 400 && (!FILTER_URL || url.includes(FILTER_URL))) {
    failures.push({ type: 'httpError', url, status });
  }
});

// Subscribe before navigating so no events are missed
await connection.subscribe(['network.fetchError', 'network.responseCompleted']);

try {
  await page.goto(TARGET_URL);

  // Allow async requests triggered after load to complete
  await new Promise(r => setTimeout(r, 2000));

  if (failures.length > 0) {
    console.log('issue reproduced:');
    for (const f of failures) {
      if (f.type === 'fetchError') console.log(`  [fetch error] ${f.url} — ${f.error}`);
      if (f.type === 'httpError')  console.log(`  [${f.status}] ${f.url}`);
    }
  } else {
    console.log('issue not reproduced');
  }
} finally {
  await session.close();
}

File: python/examples/check_network_errors.py
"""
Webcompat repro pattern: network request failures (CORS, 4xx, 5xx, fetch errors)

Use this when you observed: a failed fetch in DevTools, a CORS error in the
console, missing content that should have loaded from an API, or a red request
in the network panel.

Adjust TARGET_URL and optionally FILTER_URL to narrow to the specific endpoint.
"""

import asyncio
from firefox_bidi_client import start_session

TARGET_URL = "https://example.com"

# Optional: only report failures for URLs containing this string.
# Set to "" to report all failures.
FILTER_URL = ""


async def main():
    async with await start_session() as session:
        page, connection = session.page, session.connection

        failures = []

        def on_fetch_error(params):
            url = params["request"]["url"]
            if not FILTER_URL or FILTER_URL in url:
                failures.append({"type": "fetchError", "url": url, "error": params.get("errorText", "")})

        def on_response_completed(params):
            url = params["request"]["url"]
            status = params["response"]["status"]
            if status >= 400 and (not FILTER_URL or FILTER_URL in url):
                failures.append({"type": "httpError", "url": url, "status": status})

        connection.on_event("network.fetchError", on_fetch_error)
        connection.on_event("network.responseCompleted", on_response_completed)

        # Subscribe before navigating so no events are missed
        await connection.subscribe(["network.fetchError", "network.responseCompleted"])

        await page.goto(TARGET_URL)

        # Allow async requests triggered after load to complete
        await asyncio.sleep(2)

        if failures:
            print("issue reproduced:")
            for f in failures:
                if f["type"] == "fetchError":
                    print(f"  [fetch error] {f['url']} — {f['error']}")
                elif f["type"] == "httpError":
                    print(f"  [{f['status']}] {f['url']}")
        else:
            print("issue not reproduced")


asyncio.run(main())
